### PR TITLE
us001-view-job-roles

### PR DESCRIPTION
**Pull request by AlexMcC02**
Required functionality added with passing integration tests and tests of the service class.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
     <modelVersion>4.0.0</modelVersion>
```
Comment (shaunganley) [inline pom.xml:7]: please remove the .idea folder above and update the gitignore so this doesn't keep happening

File: pom.xml [hunk 2]
```diff
@@ -7,7 +7,7 @@
...
 
     <groupId>org.kainos.ea</groupId>
-    <artifactId>DropwizardDemo</artifactId>
+    <artifactId>TeamTwo</artifactId>
```
Comment (shaunganley) [inline pom.xml]:
```suggestion
    <artifactId>TeamToo</artifactId>
```

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+@Api("Team Too Project API")
+@Path("/api")
+public class JobRoleController {
+
+    private static JobRoleService jobRoleService;
+
+    public JobRoleController() {
+        DatabaseConnector databaseConnector = new DatabaseConnector();
+        jobRoleService = new JobRoleService(new JobRoleDao(), databaseConnector);
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java]:
```suggestion
        jobRoleService = new JobRoleService(new JobRoleDao(), new DatabaseConnector());
```

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @GET
+    @Path("/job_roles")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getAllJobRoles() {
+
+        try {
+            return Response.ok(jobRoleService.getAllJobRoles()).build();
+        } catch (FailedToGetJobRolesException | SQLException e) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java]: catch the SQLException in the service and throw a FailedToGetJobRolesException instead

File: src/main/java/org/kainos/ea/dao/JobRoleDao.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JobRoleDao {
+
+    public List<JobRole> getAllJobRoles(Connection c) throws SQLException {
+        Statement st = c.createStatement();
+
+        ResultSet rs = st.executeQuery("SELECT * FROM `JobRole`;");
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/dao/JobRoleDao.java]: only select the columns you need

File: src/main/java/org/kainos/ea/model/JobRole.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    private String specification;
+    private int BandID;
+    private int CapabilityID;
+
+    public JobRole(int id, String name, String specification, int bandID, int capabilityID) {
+        this.id = id;
+        this.name = name;
+        this.specification = specification;
+        BandID = bandID;
+        CapabilityID = capabilityID;
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/model/JobRole.java]: make these consistent with the lines above

File: src/main/java/org/kainos/ea/service/JobRoleService.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    public JobRoleDao jobRoleDao;
+    public DatabaseConnector databaseConnector;
+
+    public JobRoleService(JobRoleDao jobRoleDao, DatabaseConnector databaseConnector) {
+        this.jobRoleDao = jobRoleDao;
+        this.databaseConnector = databaseConnector;
+    }
+
+    public List<JobRole> getAllJobRoles() throws FailedToGetJobRolesException, SQLException {
+        return jobRoleDao.getAllJobRoles(databaseConnector.getConnection());
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/service/JobRoleService.java]: FailedToGetJobRolesException isn't thrown from the DAO, catch the SQLException here and throw the FailedToGetJobRolesException

File: src/main/java/org/kainos/ea/util/DatabaseConnector.java [hunk 8]
```diff
@@ -9,7 +9,7 @@
 public class DatabaseConnector {
     private static Connection conn;
 
-    public static Connection getConnection() throws SQLException {
+    public Connection getConnection() throws SQLException {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/util/DatabaseConnector.java]: why remove static?

File: src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class JobRoleIntegrationTest {
+    static final DropwizardAppExtension<DropwizardWebServiceConfiguration> APP = new DropwizardAppExtension<>(
+            DropwizardWebServiceApplication.class, null,
+            new ResourceConfigurationSourceProvider()
+    );
+
+    @Test
+    void test_1getJobRoles_shouldReturnListOfJobRoles() {
+        List<JobRole> response = APP.client().target("http://localhost:8080/api/job_roles")
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java]: don't hardcode 'http://localhost:8080' instead pass it in as an environment variable so you can set it to the AWS URL in the future

File: src/test/java/org/kainos/ea/service/JobRoleServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    JobRoleDao jobRoleDao = Mockito.mock(JobRoleDao.class);
+    DatabaseConnector databaseConnector = Mockito.mock(DatabaseConnector.class);
+
+    JobRoleService jobRoleService = new JobRoleService(jobRoleDao, databaseConnector);
+
+    Connection conn;
+
+    @Test
+    void getJobRoles_shouldReturnListOfJobRoles_whenDaoReturnsListOfJobRoles() throws SQLException, FailedToGetJobRolesException {
+        List<JobRole> expected_result = new ArrayList<>();
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/JobRoleServiceTest.java:33]: you don't use snake case anywhere else, make consistent

File: src/test/java/org/kainos/ea/service/JobRoleServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        Mockito.when(databaseConnector.getConnection()).thenReturn(conn);
+        Mockito.when(jobRoleDao.getAllJobRoles(conn)).thenReturn(expected_result);
+
+        List<JobRole> result = jobRoleService.getAllJobRoles();
+
+        assertEquals(expected_result, result);
+    }
+
+    @Test
+    void getJobRoles_shouldThrowSQLException_whenDaoThrowsSQLException() throws SQLException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/JobRoleServiceTest.java]: update this test to throw custom exception

File: src/test/java/org/kainos/ea/service/JobRoleServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+public class JobRoleServiceTest {
+
+    JobRoleDao jobRoleDao = Mockito.mock(JobRoleDao.class);
+    DatabaseConnector databaseConnector = Mockito.mock(DatabaseConnector.class);
+
+    JobRoleService jobRoleService = new JobRoleService(jobRoleDao, databaseConnector);
+
+    Connection conn;
+
+    @Test
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/JobRoleServiceTest.java:31]: add a unit test for the databaseConnector.getConnection() method throwing an exception

**Review by shaunganley — CHANGES_REQUESTED** [review]
good start. let me know if you have any questions about my comments

**Review by kamil-biegaj — COMMENTED** [review]
Checked over half of the changes, not sure if the other half has been fixed yet.

File: src/main/java/org/kainos/ea/controller/JobRoleController.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    @Path("/job_roles")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getAllJobRoles() {
+
+        try {
+            return Response.ok(jobRoleService.getAllJobRoles()).build();
+        } catch (FailedToGetJobRolesException e) {
+            System.err.println(e.getMessage());
+
+            return Response.serverError().build();
```
Comment (kamil-biegaj) [inline src/main/java/org/kainos/ea/controller/JobRoleController.java:35]: Maybe use another exception for this such as: Response.status(Response.Status.NOT_FOUND).build();

**Review by kamil-biegaj — DISMISSED** [review]

File: .gitignore [hunk 14]
```diff
@@ -41,4 +41,5 @@ build/
 db.properties
 
 ### idea folder ###
-.idea
\ No newline at end of file
+.idea
+.idea/stat.log
```
Comment (shaunganley) [inline .gitignore]:
```suggestion
.idea/
```

File: .github/workflows/maven.yml [hunk 15]
```diff
@@ -14,6 +14,12 @@ on:
   pull_request:
     branches: [ "main" ]
 
+env:
+  DB_NAME: ${{ secrets.DB_NAME }}
+  DB_USER: ${{ secrets.DB_USER }}
+  DB_PASSWORD: ${{ secrets.DB_PASSWORD }}
+  DB_HOST: ${{ secrets.DB_HOST }}
```
Comment (shaunganley) [inline .github/workflows/maven.yml]: move these under the job that needs them as they won't be required if other steps are added

File: src/test/java/org/kainos/ea/service/JobRoleServiceTest.java (hunk 10 above)
Comment (shaunganley) [reply src/test/java/org/kainos/ea/service/JobRoleServiceTest.java:33]: this is still an issue

File: src/test/java/org/kainos/ea/service/JobRoleServiceTest.java (hunk 12 above)
Comment (shaunganley) [reply src/test/java/org/kainos/ea/service/JobRoleServiceTest.java:31]: this still needs added

**Review by shaunganley — COMMENTED** [review]
getting close. please add unit tests for the controller class and address the other comments

File: src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class JobRoleIntegrationTest {
+    static final DropwizardAppExtension<DropwizardWebServiceConfiguration> APP = new DropwizardAppExtension<>(
+            DropwizardWebServiceApplication.class, null,
+            new ResourceConfigurationSourceProvider()
+    );
+
+    @Test
+    void getJobRolesShouldReturnListOfJobRoles() {
+        String url = System.getenv("DB_URL") + "/api/job_roles";
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/JobRoleIntegrationTest.java]: I don't think this is named correctly, is this an API_URL?

**Review by shaunganley — COMMENTED** [review]
great stuff, can you add some controller unit tests then this PR is good to go

**Review by philipwilson5 — COMMENTED** [review]
looks good

**Review by shaunganley — APPROVED** [review]
great work

**Review by philipwilson5 — COMMENTED** [review]
all good

**Review by philipwilson5 — APPROVED** [review]

**Review by ConnMeehan — APPROVED** [review]

**Comment by AlexMcC02** [comment]
Test

**Pull request merged**